docs/syntax.md

```diff
@@ -113,6 +113,35 @@ export let name;
 </main>
 ```
 
+### Fragments
+
+`.astro` files can render any number of top-level elements.
+
+```jsx
+---
+const ids = ['a', 'b', 'c'];
+---
+
+<div id={ids[0]} />
+<div id={ids[1]} />
+<div id={ids[2]} />
+```
+
+Inside of expressions, `.astro` can render any number of elements. There is no need for a `Fragment`-like wrapper.
+
+```jsx
+---
+const items = [0, 1, 2];
+const ids = ['a', 'b', 'c'];
+---
+
+{items.map((item, index) => (
+  <div id={`${item}-${ids[index]}`} />
+  <div id={`${item}-${ids[index]}`} />
+  <div id={`${item}-${ids[index]}`} />
+))}
+```
```

Comment:
Is this really a good idea? This isn’t consistent with JS—where multiple expressions must be comma separated, and only the last one is returned—or the closest language counterpart, JSX where multiple tags are distinct expressions. I could see this easily being a footgun. What’s the benefit of not requiring an explicit `<></>` wrapper besides a little less typing?

Reply:
Don't want to answer for @natemoo-re, but the need for a Fragment wrapper is a limitation in React/Preact, where a function must always return one thing (can't `return A B;`, so fragments act like `return [A, B]`).

Because Astro is building an HTML string, we don't have that same limitation. The analogy would be that Astro is doing `return "A B";`.

Reply:
That’s a limitation of JS/TS (a `return` statement returns only one expression) and JSX (`JSX.Element` is an expression). Until this PR, my understanding is that Astro expressions were also JS/TS, and Astro tags were also expressions.

This fundamentally reverses one or both of those, either forking JS parentheses semantics to mean some kind of list (why not just return an array then?) or turning tags into implicit template strings that may sometimes be attached to their siblings… and other times may not. In other words sometimes it’s mostly like JSX, other times some kind of JSH?

I really think from a principle of least surprise perspective, this is going to take the Astro format further from the things that make it familiar, similarly to the inversion of `export` discussed in #139.

I unfortunately can’t attend the RFC discussion tomorrow but I was really hoping to raise the possibility of a more holistic discussion around the Astro format. There are so many opportunities to use existing standards and familiar approaches rather than reinventing an uncanny valley wheel.

I’m not trying to be overly critical. I’m offering this feedback because I’m **super excited** about Astro and want to see it succeed! I just worry that the more it becomes BYOF+1, that's going add to JS ecosystem/tooling/framework fatigue and act as an impediment to adoption.

In that spirit, I’d like to reiterate the question: what is the benefit here besides saving a few keystrokes?

Reply:
@eyelidlessness Yeah I am somewhat unsure about this as well. Mostly because it's difficult to actually fix given how we handle parsing of the JS parts. But I also want to point out that we are trying to be "Mostly HTML with a little JS" and are not trying to be like JSX. There are other areas where out syntax deviates from JSX, for example we support void elements and JSX requires they be closed.

Reply:
After sleeping on this I am 100% with @eyelidlessness. This change is trying to be _too smart_ about when a user _might_ want to use a Fragment. I can see how this behavior would be unexpected and easily lead to special-casing a ton of edge cases. It would be much better to require users to be explicit. I think requiring `<>` is the most straightforward approach here.